core/lib/error.py

```diff
@@ -279,6 +279,12 @@ class OSCError(Exception):
                 '`SHOW CREATE TABLE`. Difference detected: \n {diff}'
             )
         },
+        'FAILED_TO_DECODE_DDL_FILE': {
+            'code': 150,
+            'desc': (
+                "Failed to decode DDL file '{filepath}' "
+                "with charset '{charset}'."),
```

Comment:
Would be nice to suggest users using `--charset` with correct character set, when this exception is being raised.

Reply:
Ok, sorry for my delay.
I can do both... but I see a problem in the way we are validating arguments here. We are using the same exception/error handler used for deeper, internal, exceptions/errors. And still, you want me to use the nice/user-friendly messages as if we were using the argparser error handler.
My suggestion is to have something like this instead:
```diff
diff --git a/core/commands/base.py b/core/commands/base.py
index 2af474d..1f8794d 100644
--- a/core/commands/base.py
+++ b/core/commands/base.py
@@ -95,12 +95,14 @@ class CommandBase(object):
                             help="MySQL user password to connect to the "
                             "instance",
                             required=require_password)
-        parser.add_argument("--charset",
-                            help="Character set used for MySQL connection")
+        parser.add_argument("--charset", default='ascii',
+                            help="Character set used for MySQL connection "
+                            "(defaults to '%(default)s').")
         parser.add_argument("--force",
                             help="Ignore non-critical errors and continue "
                             "making schema changes for all the given "
                             "databases ")
+        self.parser = parser
 
     def add_file_list_parser(self, parser):
         parser.add_argument("--ddl-file-list",
diff --git a/core/commands/copy.py b/core/commands/copy.py
index 7002aa7..db37e13 100644
--- a/core/commands/copy.py
+++ b/core/commands/copy.py
@@ -13,6 +13,7 @@ from __future__ import division
 from __future__ import print_function
 from __future__ import unicode_literals
 
+import argparse
 import logging
 import os
 
@@ -215,11 +216,22 @@ class Copy(CommandBase):
                 raise OSCError('OUTFILE_DIR_NOT_DIR',
                                {'dir': self.args.outfile_dir})
 
-        # Ensure all the given ddl files are readable
+        # Ensure all the given ddl files are readable and
+        # can be decoded with the current charset
         for filepath in self.args.ddl_file_list:
             if not util.is_file_readable(filepath):
                 raise OSCError('FAILED_TO_READ_DDL_FILE',
                                {'filepath': filepath})
+            charset = getattr(self.args, 'charset')
+            try:
+                with open(filepath, 'r') as file_obj:
+                    file_obj.read().decode(charset)
+            except UnicodeDecodeError:
+                raise argparse.ArgumentTypeError("Cannot decode '%s' with "
+                                                 "charset '%s'. Set the "
+                                                 "correct charset with "
+                                                 "'--charset'." %
+                                                 (filepath, charset))
 
     def op(self, sudo=False):
         self.payload = CopyPayload(get_conn_func=self.get_conn_func,
diff --git a/osc_cli b/osc_cli
index 67fcbc6..13316e7 100755
--- a/osc_cli
+++ b/osc_cli
@@ -117,6 +117,8 @@ def main():
     except OSCError as e:
         log.error(e.desc)
         sys.exit(e.code)
+    except argparse.ArgumentTypeError as details:
+        cmd.parser.error(details)
 
 
 if __name__ == '__main__':


```

Which would result in:
```
$ osc_cli copy --ddl-file-list=test2.sql --mysql-user=root --mysql-password="" --database=test --socket=/var/lib/mysql/mysql.sock
usage: osc_cli copy [-h] --socket SOCKET --database DATABASE [DATABASE ...]
                    [--repl-status {master,slave}] --mysql-user MYSQL_USER
                    --mysql-password MYSQL_PASSWORD [--charset CHARSET]
                    [--force FORCE] --ddl-file-list DDL_FILE_LIST
                    [DDL_FILE_LIST ...] [--mysql-engine MYSQL_ENGINE]
                    [--rebuild] [--skip-pk-coverage-check] [--rm-partition]
                    [--eliminate-dups] [--detailed-mismatch-info]
                    [--dump-after-checksum] [--allow-new-pk]
                    [--allow-drop-column] [--idx-recreation] [--force-cleanup]
                    [--outfile-dir OUTFILE_DIR]
                    [--pre-load-statement PRE_LOAD_STATEMENT]
                    [--post-load-statement POST_LOAD_STATEMENT]
                    [--replay-timeout REPLAY_TIMEOUT]
                    [--replay-batch-size REPLAY_BATCH_SIZE]
                    [--replay-grouping-size REPLAY_GROUPING_SIZE]
                    [--replay-max-attempt REPLAY_MAX_ATTEMPT]
                    [--reserved-space RESERVED_SPACE]
                    [--chunk-size CHUNK_SIZE] [--long-trx-time LONG_TRX_TIME]
                    [--max-running-before-ddl MAX_RUNNING_BEFORE_DDL]
                    [--ddl-guard-attempts DDL_GUARD_ATTEMPTS]
                    [--lock-max-attempts LOCK_MAX_ATTEMPTS]
                    [--mysql-session-timeout MYSQL_SESSION_TIMEOUT]
                    [--keep-tmp-table-after-exception]
                    [--session-overrides SESSION_OVERRIDES]
                    [--skip-affected-rows-check] [--skip-checksum]
                    [--skip-checksum-for-modified] [--skip-named-lock]
                    [--skip-cleanup-after-kill] [--skip-disk-space-check]
                    [--where WHERE] [--fail-for-implicit-conv]
                    [--max-wait-for-slow-query MAX_WAIT_FOR_SLOW_QUERY]
osc_cli copy: error: Cannot decode 'test2.sql' with charset 'ascii'. Set the correct charset with '--charset'.
```
Of course the solution would have to be replicated for the other argument validations.

Reply:
Anyways... let me update this PR with your requests. We can discuss that later, in a separate issue.